### PR TITLE
unicode, $ref, draft4

### PR DESCRIPTION
**Pull request by andreineculau**
Port of https://github.com/klarna/jesse/pull/37 by @emfa

---

This is a continuation on draft 4 PR.
It also contains the previous PR with reference support.

This PR adds both local and remote references. It should follow the id-attribute as far as I know. Atleast the basic cases.

### CODE REVIEW

File: src/jesse_validator_draft4.erl [hunk 1]
```diff
@@ -0,0 +1,1248 @@
...
+
+%% @doc Validate a value against a schema in a given state.
+%% Used by all combinators to run validation on a schema.
+%% @private
+validate_schema(Value, Schema, State0) ->
+  try
+    case jesse_lib:is_json_object(Schema) of
+      true ->
+        State1 = jesse_state:set_current_schema(State0, Schema),
+        State2 = jesse_schema_validator:validate_with_state(Schema, Value, State1),
```
Comment (gadgetci) [inline src/jesse_validator_draft4.erl]: According to **Elvis**:
> Line 1127 is too long:         State2 = jesse_schema_validator:validate_with_state(Schema, Value, State1),.

File: src/jesse_json_path.erl [hunk 2]
```diff
@@ -26,6 +26,18 @@
 
 -export_type([proplist/0, kvc_key/0, kvc_obj/0]).
 
+%% @doc Parse a JSON Pointer
+%% TODO: Parse according to specification. This does not accurately parse
+%%       tilde escaped JSON Pointers.
+-spec parse(JSONPointer :: string() | binary()) -> [binary()].
+parse(JSONPointer) ->
+    lists:map(
+      fun (Segment) when is_list(Segment) -> unicode:characters_to_binary(Segment);
```
Comment (gadgetci) [inline src/jesse_json_path.erl]: According to **Elvis**:
> Line 35 is too long:       fun (Segment) when is_list(Segment) -> unicode:characters_to_binary(Segment);.

File: src/jesse_state.erl [hunk 3]
```diff
@@ -138,14 +149,120 @@ set_allowed_errors(#state{} = State, AllowedErrors) ->
...
+            _:_ -> ?not_found
+          end;
+        false ->
+          ?not_found
+      end
+  end.
+
+%% @doc Resolve a new id
+%% @private
+-spec combine_id(undefined | http_uri:uri(), undefined | binary()) -> http_uri:uri().
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 228 is too long: -spec combine_id(undefined | http_uri:uri(), undefined | binary()) -> http_uri:uri()..

File: src/jesse_state.erl [hunk 4]
```diff
@@ -35,28 +35,34 @@
...
+        , resolve_reference/2
         ]).
 
 -export_type([ state/0
              ]).
 
+%% Includes
+-include("jesse_schema_validator.hrl").
+
+-define(schema_loader_fun, fun jesse_database:read/1).
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Invalid macro name schema_loader_fun on line 48. Use UPPER_CASE.

File: src/jesse_state.erl [hunk 5]
```diff
@@ -35,28 +35,34 @@
...
 -record( state
-       , { original_schema    :: jesse:json_term()
+       , { root_schema        :: jesse:json_term()
          , current_schema     :: jesse:json_term()
          , current_path       :: [binary()] %% current path in reversed order
          , allowed_errors     :: non_neg_integer() | 'infinity'
          , error_list         :: list()
          , error_handler      :: fun((#state{}) -> list() | no_return())
          , default_schema_ver :: atom()
+         , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found)
```
Comment (gadgetci) [inline src/jesse_state.erl]: According to **Elvis**:
> Line 59 is too long:          , schema_loader_fun  :: fun((binary()) -> {ok, jesse:json_term()} | jesse:json_term() | ?not_found).

**Comment by emmmile** [comment]
Looks all good to me, I would just fix the issues Elvis complains about.

**Comment by jamesaimonetti** [comment]
Thanks for reviving jesse and enhancing my original PR. I'm going to point to this repo once the $ref support is merged and kick its tires.

**Comment by andreineculau** [comment]
current status: https://github.com/for-GET/jesse/tree/merge_pr_5
looks ok despite what travis says: https://travis-ci.org/for-GET/jesse/builds/96857910

appreciate it if you can take it for a ride in your own environments and report back

**Comment by emmmile** [comment]
Hey! On my environment (R16) there is a bug that seems easy to fix (the one reported by dialyzer). But I'm not sure if this then fails with R17 or above.

I have a fork with these modifications:
https://github.com/emmmile/jesse/commit/84cf8dcbdc420b1dc7ebbddd5b7d5f808f894fe6

**Comment by andreineculau** [comment]
@emmmile the `dict:dict` fix looks awfully off, I'm afraid since R16 should go on the `ifdef(erlang_deprecated_types)` branch. Can you please double-check?

current status https://github.com/for-GET/jesse/tree/merge_pr_5 of looks all green https://travis-ci.org/for-GET/jesse/builds/97222155

**Pull request merged**